cmd/integration/commands/stages.go

```diff
@@ -843,6 +872,44 @@ func stageBodies(db kv.RwDB, ctx context.Context, logger log.Logger) error {
 	return nil
 }
 
+func stagePolygonSync(db kv.RwDB, ctx context.Context, logger log.Logger) error {
+	engine, _, stageSync, _, _ := newSync(ctx, db, nil /* miningConfig */, logger)
+	heimdallClient := engine.(*bor.Bor).HeimdallClient
+	sn, borSn, agg, _ := allSnapshots(ctx, db, logger)
+	defer sn.Close()
+	defer borSn.Close()
+	defer agg.Close()
+	br, bw := blocksIO(db, logger)
+	dirs := datadir.New(datadirCli)
+	chainConfig := fromdb.ChainConfig(db)
+
+	return db.Update(ctx, func(tx kv.RwTx) error {
+		if reset {
+			if err := reset2.ResetPolygonSync(tx, db, agg, br, bw, dirs, *chainConfig, logger); err != nil {
+				return err
+			}
+			return nil
+		}
+
+		stageState := stage(stageSync, tx, nil, stages.Senders)
+		cfg := stagedsync.NewPolygonSyncStageCfg(logger, chainConfig, nil, heimdallClient, nil, 0, nil, br, nil, 0) // we only need blockReader and blockWriter (blockWriter is constructed in NewPolygonSyncStageCfg)
+		if unwind > 0 {
+			u := stageSync.NewUnwindState(stages.Senders, stageState.BlockNumber-unwind, stageState.BlockNumber, true, false)
+			if err := stagedsync.UnwindPolygonSyncStage(ctx, tx, u, cfg); err != nil {
+				return err
+			}
+
+			for _, s := range []stages.SyncStage{stages.Snapshots, stages.Headers, stages.Bodies, stages.PolygonSync} {
```

Comment:
> I'm not sure if we should be touching the progress of the Snapshots stage for the unwind case.

I was running into an issue where `PolygonSync` continue off from the Snapshot stage block which lead to a block gap in the DB. Setting the Snapshots stage progress fixed this issue, but if you think this is incorrect then I can remove it - but we have to debug why the gap in the DB occurs.

Reply:
> Also just realised, we should be updating the Bodies and Headers progress inside the UnwindPolygonSyncStage - I've just forgotten to do it. Could you please add that there? Then this for loop here will be unnecessary I think

Most (if not all) unwind functions don't update the stage progress. I wouldn't want to do this for the sake of consistency.

Reply:
That means something does not align, have a look at `polygonSyncStageExecutionEngine.currentHeader`
There is probably a bug somewhere
Definition of unwinding at the moment is to not go back past frozen data.
I would prefer if we have an explanation what is happening - this means we need to debug and understand the problem instead of doing code changes that "make it work" without actually being able to provide an explanation of the problem.

Reply:
Why does changing stage.Snapshots progress matter at all and how does it magically fix the gap?